### PR TITLE
Elementizes and Greyscales blood decals/overlays

### PR DESCRIPTION
**Pull request by thgvr**
## About The Pull Request
This PR converts our blood decals to elements, and adds greyscale capability to blood, tied to species DNA.
Ports:
https://github.com/BeeStation/BeeStation-Hornet/pull/3111
https://github.com/BeeStation/BeeStation-Hornet/pull/3046
https://github.com/tgstation/tgstation/pull/61610
https://github.com/tgstation/tgstation/pull/59873
https://github.com/tgstation/tgstation/pull/59315
https://github.com/tgstation/tgstation/pull/53109

Some others, I don't quite remember. Mostly related fixes for those PRs.

## Why It's Good For The Game

Decals as a component is a cause of some lag, apparently. This should alleviate some if it's even noticeable.

Species having a framework for unique, greyscale blood colors can add a lot for our lore and visual feel.

## Changelog

:cl:
add: Elzu now bleed, but do not have Liquid Electricity reagent as blood. (will be changed in the future)
add: IPCs now "bleed" by leaking coolant
add: Sarathi now have teal-colored blood.
add: Blood will now dry over time
add: New bloody footprint sprites from bay
refactor: Refactors a lot of blood code
/:cl:

### CODE REVIEW

**Comment by thgvr** [comment]
Current runtime issues 
image

**Comment by thgvr** [comment]
image
Current runtimes after slight fix pr, less but still some

File: code/__HELPERS/_lists.dm [hunk 1]
```diff
@@ -634,3 +634,13 @@
 			. += sanitize_simple("[item]", list("{"="", "}"="", "\""="", ";"="", ","=""))
 		first_entry = FALSE
 	. += ")"
+
+#define LAZY_LISTS_OR(left_list, right_list)\
```
Comment (tmtmtl30) [inline code/__HELPERS/_lists.dm]: feel like this should be up near the other lazy list defines, near the top of the file

File: code/game/objects/effects/decals/cleanable/humans.dm [hunk 2]
```diff
@@ -1,17 +1,55 @@
...
+/obj/effect/decal/cleanable/blood/proc/dry()
+	if(bloodiness > 20)
+		bloodiness -= BLOOD_AMOUNT_PER_DECAL
+		get_timer()
+	else
+		name = dryname
+		desc = drydesc
+		bloodiness = 0
+		var/temp_color = ReadHSV(RGBtoHSV(color || COLOR_WHITE))
+		color = HSVtoRGB(hsv(temp_color[1], temp_color[2], max(temp_color[3] - 100, 0)))
```
Comment (tmtmtl30) [inline code/game/objects/effects/decals/cleanable/humans.dm]: might be worth changing max(temp_color[3] - 100, 0) to max(temp_color[3] - 100, min(temp_color[3], 10)), so that you don't end up with a 0-brightness blood pool unless it started that way

File: code/modules/clothing/head/_head.dm [hunk 3]
```diff
@@ -70,7 +70,10 @@
 		if(damaged_clothes)
 			. += mutable_appearance('icons/effects/item_damage.dmi', "damagedhelmet")
 		if(HAS_BLOOD_DNA(src))
-			. += mutable_appearance('icons/effects/blood.dmi', "helmetblood")
+			var/mutable_appearance/bloody_helmet
+			bloody_helmet.color = get_blood_dna_color(return_blood_DNA())
```
Comment (tmtmtl30) [inline code/modules/clothing/head/_head.dm:74]: you're modifying bloody_helmet before it's initialized (on the next line), here.

File: code/modules/clothing/suits/_suits.dm [hunk 4]
```diff
@@ -20,9 +20,11 @@
 	. = list()
 	if(!isinhands)
 		if(damaged_clothes)
-			. += mutable_appearance('icons/effects/item_damage.dmi', "damaged[blood_overlay_type]")
+			. += mutable_appearance('icons/effects/item_damage.dmi', "damageduniform")
```
Comment (tmtmtl30) [inline code/modules/clothing/suits/_suits.dm:23]: should probably still be "damaged[blood_overlay_type]" here, right?

File: code/modules/mob/living/living.dm [hunk 5]
```diff
@@ -787,9 +787,9 @@
 /mob/living/proc/makeTrail(turf/target_turf, turf/start, direction)
 	if(!has_gravity())
 		return
-	var/blood_exists = FALSE
+	var/blood_exists = locate(/obj/effect/decal/cleanable/blood/trail_holder) in start
 
-	for(var/obj/effect/decal/cleanable/trail_holder/C in start) //checks for blood splatter already on the floor
+	for(var/obj/effect/decal/cleanable/blood/trail_holder/C in start) //checks for blood splatter already on the floor
```
Comment (tmtmtl30) [inline code/modules/mob/living/living.dm]: this loop isn't necessary, since the prior line that initializes blood_exists already finds a trail_holder in start

File: code/modules/projectiles/projectile.dm [hunk 6]
```diff
@@ -236,15 +236,23 @@
 			if(starting)
 				splatter_dir = get_dir(starting, target_loca)
 			if(isalien(L))
-				new /obj/effect/temp_visual/dir_setting/bloodsplatter/xenosplatter(target_loca, splatter_dir)
+				var/splatter_color = null
+				if(iscarbon(L))
```
Comment (tmtmtl30) [inline code/modules/projectiles/projectile.dm]: don't think this is necessary; L is guaranteed to be a carbon, since it's an alien. remove this check and the contained code; modify the line below to create an /obj/effect/temp_visual/dir_setting/bloodsplatter/xenosplatter and don't pass the splatter_color (because, since it's a xeno overlay with a prebaked sprite, it doesn't need it)

**Review by tmtmtl30 — DISMISSED** [review]
a few changes

**Comment by thgvr** [comment]
Needs DNM tag removed

**Review by MarkSuckerberg — APPROVED** [review]
Seems good

**Comment by thgvr** [comment]
This was TM for quite some time. There are no outstanding issues

**Review by rye-rice — APPROVED** [review]

File: code/game/atoms.dm [hunk 7]
```diff
@@ -756,7 +756,7 @@
 	return blood_dna
 
 /mob/living/carbon/alien/get_blood_dna_list()
-	return list("UNKNOWN DNA" = "X*")
+	return list("UNKNOWN DNA" = get_blood_type("X"))
 
 /mob/living/silicon/get_blood_dna_list()
 	return list("MOTOR OIL" = "SAE 5W-30") //just a little flavor text.
```
Comment (Erikafox) [inline code/game/atoms.dm]: lxyWGp0cYz
IPC shitcode strikes again

Comment (thgvr) [reply code/game/atoms.dm]: I don't even think I can fix this

Comment (MarkSuckerberg) [reply code/game/atoms.dm]: Might get some flak for this, but honestly— just make a new blood type for IPCs. I've looked over the code, and honestly there's less work to be done that way and while it's kind of funny to use oil as IPC blood it's probably best to just make a synthetic blood type here. I'd like to ask you to remove all the code for that as well, but just adding a new IPC blood type would be just as fine. Don't forget to add the blood packs to the freezers also because. yeah

tl;dr just make a new blood type for ipcs

**Comment by goober3** [comment]
image
hi! I think this PR might be breaking stair decals.

**Comment by thgvr** [comment]
what do you mean breaking

**Comment by goober3** [comment]
> what do you mean breaking

it's making them Colorless. they should be gray.

File: code/game/atoms.dm (hunk 7 above)
Comment (MarkSuckerberg) [inline code/game/atoms.dm]:
```suggestion
	return list("SYNTHETIC COOLANT" = get_blood_type("Coolant"))
```

**Review by MarkSuckerberg — CHANGES_REQUESTED** [review]
This was runtiming a bit, easy fix

File: code/game/atoms.dm (hunk 7 above)
Comment (thgvr) [reply code/game/atoms.dm]: I was gonna do that but I forgot haha

**Review by MarkSuckerberg — APPROVED** [review]
homedstuck

**Comment by thgvr** [comment]
lets go dude

File: code/datums/blood_type.dm [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+	name = "B+"
+	compatible_types = list(/datum/blood_type/b_minus, /datum/blood_type/b_plus, /datum/blood_type/o_minus, /datum/blood_type/o_plus)
+
+/datum/blood_type/ab_minus
+	name = "AB-"
+	compatible_types = list(/datum/blood_type/b_minus, /datum/blood_type/a_minus, /datum/blood_type/ab_minus, /datum/blood_type/o_minus)
+
+/datum/blood_type/ab_plus
+	name = "AB+"
+	compatible_types = list(/datum/blood_type/b_minus, /datum/blood_type/a_minus, /datum/blood_type/ab_minus, /datum/blood_type/o_minus)
```
Comment (Ms-Mee) [inline code/datums/blood_type.dm:31]: Wha

Comment (Ms-Mee) [reply code/datums/blood_type.dm:31]: \> Be me, first day as AB+ blood bank supervisor
\> AB+ blood accepts any other kind of human blood, from O- to AB+
\> Look inside AB+ blood bank
\> It's all AB-
(I may be going insane but I don't think this is right?)

Comment (meemofcourse) [reply code/datums/blood_type.dm:31]: AB+ is an universal recipient, which includes AB- (but not the other way around)

Comment (Ms-Mee) [reply code/datums/blood_type.dm:31]: Look at the code my dearest chap

Comment (Ms-Mee) [reply code/datums/blood_type.dm:31]: image

You can't receive A+ as AB+ or even O+, but you can receive A-, AB+ has the same compatible types as AB-, without A+, B+, O+ or even itself, AB+, you take toxin damage from your own blood.

image

Yes I should open an issue but that's later me's problem.

Comment (thgvr) [reply code/datums/blood_type.dm:31]: IDK this is how it was
